src/io/fastx.rs

```diff
@@ -0,0 +1,442 @@
+use std::io;
+use std::io::prelude::*;
+use std::mem;
+
+use bio_types::sequence::SequenceRead;
+
+use crate::io::fasta;
+use crate::io::fasta::FastaRead;
+use crate::io::fastq;
+use crate::io::fastq::FastqRead;
+use crate::utils::TextSlice;
+
+const BASE_QUAL: u8 = 30;
+
+pub enum Reader<R: io::Read> {
+    FASTA(fasta::Reader<R>),
+    FASTQ(fastq::Reader<R>),
+    Uninitialized(Option<R>),
+}
+
+pub enum Record {
+    FASTA(fasta::Record),
+    FASTQ(fastq::Record),
+}
+
+pub struct Records<R: io::Read> {
+    reader: Reader<R>,
+    error_has_occured: bool,
+}
+
+impl Record {
+    pub fn new() -> Self {
+        Record::FASTA(fasta::Record::new())
+    }
+
+    pub fn with_attrs(
+        id: &str,
+        desc: Option<&str>,
+        seq: TextSlice<'_>,
+        qual: Option<&[u8]>,
+    ) -> Self {
+        match qual {
+            Some(q) => Record::FASTQ(fastq::Record::with_attrs(id, desc, seq, q)),
+            None => Record::FASTA(fasta::Record::with_attrs(id, desc, seq)),
+        }
+    }
+
+    pub fn id(&self) -> &str {
+        match self {
+            Record::FASTA(x) => x.id(),
+            Record::FASTQ(x) => x.id(),
+        }
+    }
+
+    pub fn check(&self) -> Result<(), &str> {
+        match self {
+            Record::FASTA(x) => x.check(),
+            Record::FASTQ(x) => x.check(),
+        }
+    }
+
+    pub fn desc(&self) -> Option<&str> {
+        match self {
+            Record::FASTA(x) => x.desc(),
+            Record::FASTQ(x) => x.desc(),
+        }
+    }
+
+    pub fn seq(&self) -> TextSlice<'_> {
+        match self {
+            Record::FASTA(x) => x.seq(),
+            Record::FASTQ(x) => x.seq(),
+        }
+    }
+
+    pub fn qual(&self) -> Vec<u8> {
```

Comment:
Would it be possible to return `Option<Vec<u8>>` instead (`None` for FASTA, similar to `desc(&self)`)? If I understood this right, right now a vector with dummy quality values is generated when `qual` is called. I think trying to get quality values for FASTA reads should just not work to avoid this allocation.

Reply:
That makes a lot of sense! I was trying to mirror the `fastq::Record` methods, but seems more reasonable to let whoever is calling the method to check if there is a quality value and then adapt as needed.